sdlib/dmd/thread.d

```diff
@@ -19,16 +20,17 @@ void __sd_scanAllThreadsFn(ScanDg* context, void* start, void* end) {
 	(*context)(makeRange(start, end));
 }
 
-// defined in druntime. The context pointer gets passed to the scan routine
-void thread_scanAll_C(ScanDg* context, typeof(&__sd_scanAllThreadsFn) scan);
-
 // sdrt API.
 void __sd_thread_scan(ScanDg scan) {
-	import d.rt.stack;
-	__sd_stack_scan(scan);
+	// When druntime is being used, all thread scanning is done by
+	// thread_scanAll_C, and not via SDC's thread scanning.
```

Comment:
Problem is, scan all will:
1. Push the register oen stack.
2. Put the stack on the worklist.
3. Pop the register values.
4. Scan the stack.

So we might be missing roots. On the other hand, we'd destroy any level of parallelism we have if we scan all the ranges as the are passed in by scanAll.

So we can't really do away with scanning that stack because DMD's API is a little dum dum.

Reply:
I think either you are misunderstanding what is happening here, or I am missing it.

What I think happens is that the GC scanner [spawns CPUcount threads](https://github.com/snazzy-d/sdc/blob/3bda7b0b505ea25f9858eb48262ecceb600a72a9/sdlib/d/gc/scanner.d#L62) to help work on the scans. Then each thread first calls this function, which scans the stack for this *newly created thread*, which obviously has no live things in it (see [here](https://github.com/snazzy-d/sdc/blob/3bda7b0b505ea25f9858eb48262ecceb600a72a9/sdlib/d/gc/scanner.d#L101)), and finally starts pulling off work to scan from the list (the while loop after that).

Aside from the problem there (which I think is not a DMD-specific problem), Druntime does not work like this, the `thread_suspendAll` function does all the things for the live threads by [putting the registers on the stack](https://github.com/dlang/dmd/blob/6835faefde3cde3deb30247ca38edf163e88f0c2/druntime/src/core/thread/osthread.d#L1547-L1553), and then storing in each thread object the stack that needs scanning (including the registers).

So even if this was fixed to be called from threads that are live threads (and not newly spawned helper threads), druntime does not need this functionality.

Reply:
Note that `thread_scanAll_C` just adds work (TLS + stack of each thread) to the list, it doesn't actually scan the stacks. The scanning is still done by the worker threads.

Reply:
One of these thread is going to be the thread that called collect.

Reply:
Yes, and that thread's info is also captured with `thread_suspendAll`.

For sure, since SDC doesn't have a stop-the-world implementation, how this should work is undecided. But there is zero reason to scan the newly-spawned work thread stacks.

Reply:
As discussed over slack, I agree with you, I added a note on why we keep it there, so future me doesn't question it...